cmd/skopeo/completions.go

```diff
@@ -1,13 +1,57 @@
 package main
 
 import (
+	"github.com/containers/image/v5/directory"
+	"github.com/containers/image/v5/docker"
+	dockerArchive "github.com/containers/image/v5/docker/archive"
+	ociArchive "github.com/containers/image/v5/oci/archive"
+	oci "github.com/containers/image/v5/oci/layout"
+	"github.com/containers/image/v5/sif"
 	"github.com/containers/image/v5/tarball"
 	"github.com/containers/image/v5/transports"
 	"github.com/spf13/cobra"
+	"os"
+	"path"
+	"path/filepath"
+	"strings"
 )
 
-// autocompleteSupportedTransports list all supported transports with the colon suffix.
-func autocompleteSupportedTransports(cmd *cobra.Command, args []string, toComplete string) ([]cobra.Completion, cobra.ShellCompDirective) {
+func autocompleteTransports(cmd *cobra.Command, args []string, toComplete string) ([]cobra.Completion, cobra.ShellCompDirective) {
+	directive := cobra.ShellCompDirectiveNoSpace
+	// We still don't have a transport
+	transport, details, haveTransport := strings.Cut(toComplete, ":")
+	if !haveTransport {
+		transports := supportedTransportSuggestions()
+		return transports, directive | cobra.ShellCompDirectiveNoFileComp
+	}
+	if transport == ociArchive.Transport.Name() || transport == dockerArchive.Transport.Name() || transport == sif.Transport.Name() || transport == oci.Transport.Name() {
+		return nil, directive
+	}
+	if transport == docker.Transport.Name() && details == "" {
+		return []cobra.Completion{docker.Transport.Name() + "://"}, directive | cobra.ShellCompDirectiveNoFileComp
+	}
+	if transport == directory.Transport.Name() {
+		// just ShellCompDirectiveFilterDirs is more correct, but doesn't work here in bash, see https://github.com/spf13/cobra/issues/2242. Instead we get the directories ourselves.
+		curDir := filepath.Dir(details)
+		entries, err := os.ReadDir(curDir)
+		if err != nil {
+			cobra.CompErrorln("Failed ReadDir at " + curDir)
+			// Fallback to whatever the shell gives us
+			return nil, cobra.ShellCompDirectiveFilterDirs | cobra.ShellCompDirectiveNoFileComp
```

Comment:
The way I read the documentation, `NoFileComp` should make `FilterDirs` entirely irrelevant. (In the implementation, that seems not to matter.)

Reply:
Yeah ShellCompDirectiveFilterDirs by itself should just complete directories (though in bash it currently doesn't). fixed